File: crates/kuben-agent/src/tls.rs
//! TLS for AgentLink (ADR-027, from the M0 spike): TLS 1.3 only, the `ring`
//! provider, mutual authentication.
//!
//! * The agent trusts only the pinned hub CA and checks the hub's name.
//! * The hub accepts client certificates issued by the cluster CA. With
//!   [`ClientAuth::EnrollmentAllowed`] it also accepts a connection without
//!   one: such a peer is anonymous and may only enroll.
//! * Extended key usage keeps the roles apart: a server certificate is never
//!   an agent identity.

use std::sync::Arc;

use rustls::{
    ClientConfig, RootCertStore, ServerConfig, ServerConnection,
    crypto::{CryptoProvider, ring},
    pki_types::{CertificateDer, PrivateKeyDer, ServerName},
    server::WebPkiClientVerifier,
    version::TLS13,
};

/// The name the hub's certificate carries unless configured otherwise.
pub const HUB_NAME: &str = "hub.kuben.internal";

/// A certificate chain and its private key.
#[derive(Debug)]
pub struct Identity {
    pub chain: Vec<CertificateDer<'static>>,
    pub key: PrivateKeyDer<'static>,
}

impl Clone for Identity {
    fn clone(&self) -> Self {
        Self {
            chain: self.chain.clone(),
            key: self.key.clone_key(),
        }
    }
}

/// Whether the hub lets a peer in without a client certificate.
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum ClientAuth {
    /// Every peer presents a certificate issued by the cluster CA.
    Required,
    /// A peer without a certificate gets in too, anonymous: it may only
    /// enroll.
    EnrollmentAllowed,
}

#[derive(Debug, thiserror::Error)]
pub enum TlsError {
    #[error("TLS: {0}")]
    Rustls(#[from] rustls::Error),
    #[error("client certificate verifier: {0}")]
    Verifier(#[from] rustls::server::VerifierBuilderError),
    #[error("no trusted CA certificate was given")]
    NoTrust,
    #[error("`{0}` is not a valid server name")]
    ServerName(String),
}

/// The crypto provider of both ends.
#[must_use]
pub fn provider() -> Arc<CryptoProvider> {
    Arc::new(ring::default_provider())
}

fn roots(cas: &[CertificateDer<'static>]) -> Result<Arc<RootCertStore>, TlsError> {
    if cas.is_empty() {
        return Err(TlsError::NoTrust);
    }
    let mut store = RootCertStore::empty();
    for ca in cas {
        store.add(ca.clone())?;
    }
    Ok(Arc::new(store))
}

/// The agent's end: trusts `pinned` alone for the hub, and authenticates
/// with `identity` once enrolled (without one, it can only enroll).
pub fn agent_config(
    pinned: &[CertificateDer<'static>],
    identity: Option<Identity>,
) -> Result<Arc<ClientConfig>, TlsError> {
    let builder = ClientConfig::builder_with_provider(provider())
        .with_protocol_versions(&[&TLS13])?
        .with_root_certificates(roots(pinned)?);
    Ok(Arc::new(match identity {
        Some(id) => builder.with_client_auth_cert(id.chain, id.key)?,
        None => builder.with_no_client_auth(),
    }))
}

/// The hub's end: presents `identity`, and accepts client certificates
/// issued by `cluster_cas` (or, with [`ClientAuth::EnrollmentAllowed`], no
/// certificate at all).
pub fn hub_config(
    cluster_cas: &[CertificateDer<'static>],
    identity: Identity,
    auth: ClientAuth,
) -> Result<Arc<ServerConfig>, TlsError> {
    let verifier = WebPkiClientVerifier::builder_with_provider(roots(cluster_cas)?, provider());
    let verifier = match auth {
        ClientAuth::Required => verifier,
        ClientAuth::EnrollmentAllowed => verifier.allow_unauthenticated(),
    }
    .build()?;
    Ok(Arc::new(
        ServerConfig::builder_with_provider(provider())
            .with_protocol_versions(&[&TLS13])?
            .with_client_cert_verifier(verifier)
            .with_single_cert(identity.chain, identity.key)?,
    ))
}

/// The hub's name as the agent checks it.
pub fn server_name(name: &str) -> Result<ServerName<'static>, TlsError> {
    ServerName::try_from(name.to_owned()).map_err(|_| TlsError::ServerName(name.to_owned()))
}

/// The certificate the peer authenticated with; `None` for an anonymous
/// peer (hub side).
#[must_use]
pub fn peer_certificate(conn: &ServerConnection) -> Option<CertificateDer<'static>> {
    conn.peer_certificates()
        .and_then(|chain| chain.first())
        .map(|c| c.clone().into_owned())
}

#[cfg(test)]
mod tests {
    use rcgen::{
        BasicConstraints, CertificateParams, DnType, ExtendedKeyUsagePurpose, IsCa, Issuer, KeyPair,
        KeyUsagePurpose,
    };
    use rustls::pki_types::PrivatePkcs8KeyDer;
    use tokio::io::{AsyncReadExt, AsyncWriteExt, duplex};
    use tokio_rustls::{TlsAcceptor, TlsConnector};

    use super::*;

    struct Ca {
        issuer: Issuer<'static, KeyPair>,
        der: CertificateDer<'static>,
    }

    fn ca(name: &str) -> Ca {
        let mut params = CertificateParams::new(Vec::<String>::new()).expect("ca params");
        params.is_ca = IsCa::Ca(BasicConstraints::Unconstrained);
        params.distinguished_name.push(DnType::CommonName, name);
        params.key_usages = vec![KeyUsagePurpose::KeyCertSign, KeyUsagePurpose::CrlSign];
        let key = KeyPair::generate().expect("ca key");
        let der = params.self_signed(&key).expect("self-signed").der().clone();
        Ca {
            issuer: Issuer::new(params, key),
            der,
        }
    }

    fn leaf(ca: &Ca, common_name: &str, sans: &[&str], purpose: ExtendedKeyUsagePurpose) -> Identity {
        let sans = sans.iter().map(|s| (*s).to_owned()).collect::<Vec<_>>();
        let mut params = CertificateParams::new(sans).expect("leaf params");
        params.distinguished_name.push(DnType::CommonName, common_name);
        params.extended_key_usages = vec![purpose];
        let key = KeyPair::generate().expect("leaf key");
        let cert = params.signed_by(&key, &ca.issuer).expect("sign");
        Identity {
            chain: vec![cert.der().clone()],
            key: PrivateKeyDer::Pkcs8(PrivatePkcs8KeyDer::from(key.serialize_der())),
        }
    }

    fn hub_identity(ca: &Ca) -> Identity {
        leaf(ca, "kuben-hub", &[HUB_NAME], ExtendedKeyUsagePurpose::ServerAuth)
    }

    fn agent_identity(ca: &Ca) -> Identity {
        leaf(ca, "cluster:primary", &[], ExtendedKeyUsagePurpose::ClientAuth)
    }

    /// One round trip; the certificate the hub saw (if any) when both ends
    /// accepted each other.
    async fn exchange(
        hub: Arc<ServerConfig>,
        agent: Arc<ClientConfig>,
    ) -> Result<Option<CertificateDer<'static>>, String> {
        let (agent_io, hub_io) = duplex(64 * 1024);
        let hub_side = tokio::spawn(async move {
            let mut tls = TlsAcceptor::from(hub)
                .accept(hub_io)
                .await
                .map_err(|e| format!("hub: {e}"))?;
            let peer = peer_certificate(tls.get_ref().1);
            let mut hello = [0u8; 5];
            tls.read_exact(&mut hello)
                .await
                .map_err(|e| format!("hub read: {e}"))?;
            tls.write_all(b"pong")
                .await
                .map_err(|e| format!("hub write: {e}"))?;
            tls.shutdown().await.map_err(|e| format!("hub shutdown: {e}"))?;
            Ok::<_, String>(peer)
        });
        let agent_side = async {
            let mut tls = TlsConnector::from(agent)
                .connect(server_name(HUB_NAME).map_err(|e| e.to_string())?, agent_io)
                .await
                .map_err(|e| format!("agent: {e}"))?;
            tls.write_all(b"hello")
                .await
                .map_err(|e| format!("agent write: {e}"))?;
            let mut reply = Vec::new();
            tls.read_to_end(&mut reply)
                .await
                .map_err(|e| format!("agent read: {e}"))?;
            Ok::<_, String>(reply)
        }
        .await;
        let hub_result = hub_side.await.map_err(|e| e.to_string())?;
        match (agent_side, hub_result) {
            (Ok(reply), Ok(peer)) if reply == b"pong" => Ok(peer),
            (Ok(reply), Ok(_)) => Err(format!("unexpected reply {reply:?}")),
            (Err(e), _) | (_, Err(e)) => Err(e),
        }
    }

    fn hub(ca: &Ca, identity: Identity, auth: ClientAuth) -> Arc<ServerConfig> {
        hub_config(std::slice::from_ref(&ca.der), identity, auth).expect("hub config")
    }

    fn agent(pinned: &Ca, identity: Option<Identity>) -> Arc<ClientConfig> {
        agent_config(std::slice::from_ref(&pinned.der), identity).expect("agent config")
    }

    #[tokio::test]
    async fn an_enrolled_agent_and_the_pinned_hub_authenticate_each_other() {
        let root = ca("kuben CA");
        let identity = agent_identity(&root);
        let cert = identity.chain[0].clone();
        let seen = exchange(
            hub(&root, hub_identity(&root), ClientAuth::Required),
            agent(&root, Some(identity)),
        )
        .await
        .expect("mutual TLS");
        assert_eq!(
            seen,
            Some(cert),
            "the hub identifies the agent by its certificate"
        );
    }

    #[tokio::test]
    async fn a_client_certificate_from_another_ca_is_rejected() {
        let root = ca("kuben CA");
        let rogue = ca("rogue CA");
        let result = exchange(
            hub(&root, hub_identity(&root), ClientAuth::EnrollmentAllowed),
            agent(&root, Some(agent_identity(&rogue))),
        )
        .await;
        assert!(result.is_err(), "{result:?}");
    }

    #[tokio::test]
    async fn an_anonymous_peer_gets_in_only_where_enrollment_is_allowed() {
        let root = ca("kuben CA");
        let refused = exchange(
            hub(&root, hub_identity(&root), ClientAuth::Required),
            agent(&root, None),
        )
        .await;
        assert!(refused.is_err(), "{refused:?}");
        let anonymous = exchange(
            hub(&root, hub_identity(&root), ClientAuth::EnrollmentAllowed),
            agent(&root, None),
        )
        .await
        .expect("an anonymous peer may enroll");
        assert_eq!(anonymous, None, "and is known to be anonymous");
    }

    #[tokio::test]
    async fn the_agent_refuses_a_hub_it_has_not_pinned_or_of_another_name() {
        let root = ca("kuben CA");
        let impostor = ca("impostor CA");
        let unpinned = exchange(
            hub(&root, hub_identity(&impostor), ClientAuth::Required),
            agent(&root, Some(agent_identity(&root))),
        )
        .await;
        assert!(unpinned.is_err(), "{unpinned:?}");
        let other_name = leaf(
            &root,
            "kuben-hub",
            &["other.example.com"],
            ExtendedKeyUsagePurpose::ServerAuth,
        );
        let renamed = exchange(
            hub(&root, other_name, ClientAuth::Required),
            agent(&root, Some(agent_identity(&root))),
        )
        .await;
        assert!(renamed.is_err(), "{renamed:?}");
    }

    #[tokio::test]
    async fn a_server_certificate_is_not_an_agent_identity() {
        let root = ca("kuben CA");
        let result = exchange(
            hub(&root, hub_identity(&root), ClientAuth::Required),
            agent(&root, Some(hub_identity(&root))),
        )
        .await;
        assert!(result.is_err(), "{result:?}");
    }

    #[test]
    fn a_configuration_without_a_trusted_ca_is_refused() {
        assert!(matches!(agent_config(&[], None), Err(TlsError::NoTrust)));
        assert!(matches!(server_name("not a name"), Err(TlsError::ServerName(_))));
    }
}

File: go/kubenapi/protocol/tls.go
package protocol

// TLS for AgentLink (ADR-027, from the M0 spike; tls.rs): TLS 1.3 only,
// mutual authentication.
//
//   - The agent trusts only the pinned hub CA and checks the hub's name.
//   - The hub accepts client certificates issued by the cluster CA. With
//     [ClientAuthEnrollmentAllowed] it also accepts a connection without
//     one: such a peer is anonymous and may only enroll.
//   - Extended key usage keeps the roles apart: a server certificate is
//     never an agent identity (crypto/x509 checks clientAuth on the hub and
//     serverAuth on the agent, as webpki did).

import (
	"crypto/tls"
	"crypto/x509"
	"errors"
	"fmt"
	"net"
	"strings"
)

// HubName is the name the hub's certificate carries unless configured
// otherwise.
const HubName = "hub.kuben.internal"

// ClientAuth is whether the hub lets a peer in without a client
// certificate.
type ClientAuth int

// The client authentication modes.
const (
	// ClientAuthRequired: every peer presents a certificate issued by the
	// cluster CA.
	ClientAuthRequired ClientAuth = iota
	// ClientAuthEnrollmentAllowed: a peer without a certificate gets in too,
	// anonymous: it may only enroll.
	ClientAuthEnrollmentAllowed
)

// ErrNoTrust is a configuration without a trusted CA certificate.
var ErrNoTrust = errors.New("no trusted CA certificate was given") //nolint:gochecknoglobals // a sentinel error

// ServerNameError is a hub name that is not a valid DNS name or IP address.
type ServerNameError struct{ Name string }

func (e ServerNameError) Error() string {
	return fmt.Sprintf("`%s` is not a valid server name", e.Name)
}

// ServerName checks that name is a DNS name or an IP address the agent can
// check the hub's certificate against (rustls' ServerName::try_from).
func ServerName(name string) (string, error) {
	if net.ParseIP(name) != nil || validDNSName(name) {
		return name, nil
	}
	return "", ServerNameError{Name: name}
}

// validDNSName follows rustls-pki-types' DnsName rules: labels of ASCII
// letters, digits, `-` and `_`, 1 to 63 bytes, not starting or ending with
// `-`, at most 253 bytes, an optional final dot, and a last label that is
// not all digits.
func validDNSName(name string) bool {
	name = strings.TrimSuffix(name, ".")
	if name == "" || len(name) > 253 {
		return false
	}
	allNumeric := true
	for label := range strings.SplitSeq(name, ".") {
		if label == "" || len(label) > 63 || label[0] == '-' || label[len(label)-1] == '-' {
			return false
		}
		allNumeric = true
		for i := range len(label) {
			c := label[i]
			switch {
			case c >= '0' && c <= '9':
			case c >= 'a' && c <= 'z', c >= 'A' && c <= 'Z', c == '-', c == '_':
				allNumeric = false
			default:
				return false
			}
		}
	}
	return !allNumeric
}

func roots(cas []*x509.Certificate) (*x509.CertPool, error) {
	if len(cas) == 0 {
		return nil, ErrNoTrust
	}
	pool := x509.NewCertPool()
	for _, ca := range cas {
		pool.AddCert(ca)
	}
	return pool, nil
}

// AgentConfig is the agent's end: it trusts pinned alone for the hub named
// hubName, and authenticates with identity once enrolled (without one, it
// can only enroll).
func AgentConfig(pinned []*x509.Certificate, hubName string, identity *tls.Certificate) (*tls.Config, error) {
	pool, err := roots(pinned)
	if err != nil {
		return nil, err
	}
	name, err := ServerName(hubName)
	if err != nil {
		return nil, err
	}
	cfg := &tls.Config{
		MinVersion: tls.VersionTLS13,
		MaxVersion: tls.VersionTLS13,
		RootCAs:    pool,
		ServerName: name,
	}
	if identity != nil {
		chosen := *identity
		// Always the one identity, as rustls sent it: crypto/tls would send
		// none when the hub's CA hints do not match, and a certificate the
		// hub refuses must be refused, not turned into an anonymous peer.
		cfg.GetClientCertificate = func(*tls.CertificateRequestInfo) (*tls.Certificate, error) {
			return &chosen, nil
		}
	}
	return cfg, nil
}

// HubConfig is the hub's end: it presents identity, and accepts client
// certificates issued by clusterCAs (or, with ClientAuthEnrollmentAllowed,
// no certificate at all).
func HubConfig(clusterCAs []*x509.Certificate, identity tls.Certificate, auth ClientAuth) (*tls.Config, error) {
	pool, err := roots(clusterCAs)
	if err != nil {
		return nil, err
	}
	if len(identity.Certificate) == 0 || identity.PrivateKey == nil {
		return nil, errors.New("TLS: the hub has no certificate")
	}
	mode := tls.RequireAndVerifyClientCert
	switch auth {
	case ClientAuthRequired:
	case ClientAuthEnrollmentAllowed:
		mode = tls.VerifyClientCertIfGiven
	}
	return &tls.Config{
		MinVersion:   tls.VersionTLS13,
		MaxVersion:   tls.VersionTLS13,
		Certificates: []tls.Certificate{identity},
		ClientCAs:    pool,
		ClientAuth:   mode,
	}, nil
}

// PeerCertificate is the certificate the peer authenticated with; false
// for an anonymous peer (hub side).
func PeerCertificate(state tls.ConnectionState) (*x509.Certificate, bool) {
	if len(state.PeerCertificates) == 0 {
		return nil, false
	}
	return state.PeerCertificates[0], true
}
